Blog Post: Vulnerable Microservices (#38104)
kubernetes/website: content/en/blog/_posts/2022-11-27-Vulnerable-Microservices/index.md

---
layout: blog
title: "All Microservices Are Vulnerable"
date: 2022-11-27
slug: security-behavior
---

**Author:**
David Hadas (IBM Research Labs)

As cyber war games continue to intensify in sophistication, organizations deploying cloud services, continue to grow their cyber investments aiming to produce safe and non-vulnerable services. However, the year-by-year growth in cyber investments, does not result in parallel reduction in cyber incidents. Instead, the number of cyber incidents continue to grow annually. Evidently organizations are fighting a battle they are doomed to fail at. No matter how much effort is made to detect and remove cyber weaknesses from deployed services, it seems offenders have the upper hand.

Considering the current spread of offensive tools, sophistication of offensive players, and ever-growing cyber financial gains to offenders, any cyber strategy that relies on constructing a non-vulnerable, weakness-free service in 2022 is clearly too naïve. It seems the only viable strategy is to:

> ***
> 
> <h3 align="center">
> Admit that your services are vulnerable
> </h3>
> 
> ***

Reviewer: ```suggestion
**Admit that your services are vulnerable**
```

We'll only accept blog articles that follow our style conventions. If that's a deal-breaker, we can discuss this - but I don't think the blog team are likely to budge.

Author: ok, let me try another approach - I am trying to ensure this sentence properly stick-out, I will send s revised option, hopefully, one that can be accepted.

Next,let's discuss why microservice architecture is well suited to security-behavior monitoring.

## Security-Behavior of Microservices vs Monoliths
Kubernetes is often used to support workloads designed with microservices architecture. By design, microservices aim to follow the UNIX philosophy of "Do One Thing And Do It Well". Each microservice has a bounded context and a clear interface. In other words, we can expect the microservice clients to send relatively regular requests and the microservice to present a relatively regular behavior as a response to these requests. Consequently, microservice architecture is an excellent candidate for security-behavior monitoring.

{{< figure src="Microservices.svg" alt="Image showing why microservices are well suited for security-behavior monitoring" class="diagram-large" caption="Figure 2. Microservices are well suited for security-behavior monitoring" >}}

> The diagram above clarifies how dividing a monolithic service to a set of microservices improves our ability to perform security-behavior monitoring and control. In a monolithic service approach, different client requests intertwined, resulting in diminished ability to identity irregular client behaviors. Without prior knowledge, an observer of the intertwined client requests will find it hard to distinguish between types of requests and their related characteristics. Further, internal client requests are not exposed to the observer. Lastly, the aggregated behavior of the monolithic service is a compound of the many different internal behaviors of its components, making it hard to identify irregular service behavior. 

Reviewer 2: ```suggestion
The diagram above clarifies how dividing a monolithic service to a set of microservices improves our ability to perform security-behavior monitoring and control. In a monolithic service approach, different client requests intertwined, resulting in diminished ability to identify irregular client behaviors. Without prior knowledge, an observer of the intertwined client requests will find it hard to distinguish between types of requests and their related characteristics. Further, internal client requests are not exposed to the observer. Lastly, the aggregated behavior of the monolithic service is a compound of the many different internal behaviors of its components, making it hard to identify irregular service behavior. 
```

---
layout: blog

> title: "Vulnerable Microservices"

Reviewer: This looks like we're reporting a vulnerability in Kubernetes, or might do. Please tweak the title to avoid that.

date: 2022-11-27
slug: vulnerable-microservices
---

**Author:** 
David Hadas (IBM Research Labs)

As cyber war games continue to intensify in sophistication, organizations deploying cloud services, continue to grow their cyber investments aiming to produce safe and non-vulnerable services. However, the year-by-year growth in cyber investments, does not result in parallel reduction in cyber incidents. Instead, the number of cyber incidents continue to grow annually. Evidently organizations are fighting a battle they are doomed to fail at. No matter how much effort is made to detect and remove cyber weaknesses from deployed services, it seems offenders have the upper hand.

> Considering the current spread of offensive tools, sophistication of offensive players, and ever-growing cyber financial gains to offenders, any cyber strategy relying on constructing a non-vulnerable, weakness-free service in 202x is clearly too naïve. It seems the only viable strategy is to:

Reviewer: ```suggestion
Considering the current spread of offensive tools, sophistication of offensive players, and ever-growing cyber financial gains to offenders, any cyber strategy that relies on constructing a non-vulnerable, weakness-free service in 2022 is clearly too naïve. It seems the only viable strategy is to:
```

> ***
> <h3 align="center">
> Admit your services are vulnerable
> </h3>
> 
> ***

Reviewer: ```suggestion
_Admit that your services may be vulnerable._
```

Only use HTML if there's a really good case for it. This doesn't feel like one.

Author: I did nit find a new way to get this test standout as it shoyld. Unless tgere are other thoughts, I will revert back to the original HTML.

Reviewer: Please see https://github.com/kubernetes/website/pull/38104#pullrequestreview-1202795661

In other words, consciously accept that you will lose any battle aimed to create non-vulnerable services. If your opponents find a single weakness as an entry-point, you lose. Admitting that although your best efforts, all your services are most likely vulnerable is an important first step. Next, we need to discuss what can you do about it...

> _This post focuses on microservices deployed with Kubernetes and points to [Guard](http://knative.dev/security-guard), an open source project designed to help monitor and control your Kubernetes deployed microservices, given that they are vulnerable._  

Author: This post is about cyber challenges that Kubernetes users have and is not aimed at Knative users. It discusses the problem users face and methods to solve it. Lastly, it points to a Knative subproject that was developed to solve this problem in Knative and indicates that it is suitable as a starting point for solving Kubernetes use cases already today and we as a community should grow it to cover additional Kubernetes use cases.

## Lose the battle - win the war!
Being vulnerable does not necessarily mean you lost the war. Though your services are vulnerable in some ways unknown to you, offenders still need to identify these vulnerabilities and then exploit them to win the war. If offenders fail to exploit your service vulnerabilities, you win! In other words, [having a vulnerability that can’t be exploited, represents a risk that can’t be realized](https://davidhadas.wordpress.com/2021/02/12/threat-vs-risk-vs-vulnerability/).

<p align="center">
<img alt="Image of an example of ofender gaining foothold in a service" width="90%" src="Example.png">

> </p>

Author: changed to figure with png 
- Btw, why should we not be using `img`? 
- How can I center the diagram with a `figure`? should I leave the `<p align="center">`?

> </p>

Author: Moved to svg - still need to review the page about the  shortcode more carfully to see if there is more to be done here..

Kubernetes deployments seeking to add Security-Behavior may use [Guard](http://knative.dev/security-guard), developed under the CNCF project Knative. **Guard can be installed to protect any HTTP-based Kubernetes service** and is <u>also</u> well integrated into the Knative automation suite that runs on top of Kubernetes.

See:
- [Guard on Github](https://github.com/knative-sandbox/security-guard)
- The Knative automation suite - Read about Knative, as an [Opinionated Kubernetes](https://davidhadas.wordpress.com/2022/08/29/knative-an-opinionated-kubernetes) implementation that simplifies and unifies the way web services are deployed on Kubernetes.
- You can also find more information on how to [secure microservices by monitoring behavior](https://developer.ibm.com/articles/secure-microservices-by-monitoring-behavior/)

<i>
As a CNCF project, Knative and a whole, and Guard in specific, is a welcoming and open community. Guard is a new and extendable project that can and should be enhanced to:

1. Cover additional kubernetes use cases,
1. Improve its monitoring and detection capabilities,
1. Improve its machine learning capabilities, and
1. Improve its integration into kubernetes.

> </i>

Reviewer: Please write this in plain Markdown.
